File: engine/language_client_typescript/typescript_src/checked.ts
export interface Checked<T,CheckName extends string = string> {
    value: T,
    checks: Record<CheckName, Check>,
}

export interface Check {
    name: string,
    expr: string
    status: "succeeded" | "failed"
}

export function all_succeeded<CheckName extends string>(checks: Record<CheckName, Check>): boolean {
    return get_checks(checks).every(check => check.status === "succeeded")
}

export function get_checks<CheckName extends string>(checks: Record<CheckName, Check>): Check[] {
    return Object.values(checks)
}

File: engine/language_client_ruby/lib/checked.rb
require "sorbet-runtime"

module Baml
  module Checks

    class Check < T::Struct
      extend T::Sig

      const :name, String
      const :expr, String
      const :status, String


      def initialize(props)
        super(name: props[:name], expr: props[:expr], status: props[:status])
      end
    end

    class Checked < T::Struct
      extend T::Sig

      extend T::Generic

      Value = type_member

      const :value, Value
      const :checks, T::Hash[Symbol, Check]

      def initialize(props)
        super(value: props[:value], checks: props[:checks])
      end

    end

  end
end
